test/common/cdp.py

```diff
@@ -38,6 +37,13 @@ def browser_path(headless=True):
             return p
         raise
 
+def have_browser():
+    try:
+        if browser_path():
+            return True
+    except Exception:
```

Comment:
No blanket exceptions please, as they also potentially hide important/stupid stuff like `NameError` or even `SyntaxError`. The relevant thing here should be `subprocess.CalledProcessError`, or you change `browser_path` to just return `None` and drop exception handling here altogether. Then you can also drop this helper function entirely.

Reply:
I think I want to shuffle this code around a bit more, in fact: as it stands right now, this problem manifests itself as an exception thrown when calling .invoke() on the cdp object, which is done after a lot of setup is already performed in tap-cdp.  This means that if you wait for this, you have to do a lot of teardown as well, otherwise stderr stays open and the tap driver stalls...  That's why the order is a bit awkward in tap-cdp right now (and why we actually shell out twice).

So I'm thinking instead, we could create the CDP object and call a .browser_exists() on it, allowing us to save the browser path in the CDP object.  This could be done before spawning the server, and before invoke.

As for the exceptions: I guess nobody is expected to call the browser path function from externally, so that's not really API, right?  But the failure mode when no browser exists (ie: .invoke() throws a subprocess error) *is* sort of like API....

Thoughts?

Reply:
`browser_path()` is not (meant to be) API indeed, and I'm not aware of any external project that uses it. So it's fine to get changed to return `None` instead of a `CalledSubprocessError` if chromium doesn't exist.

Wrt. "drop this helper function", I just meant the new `browser_exists()`. It *is* of course useful to check `if not browser_path():` right at the start, before the heavy initialization starts indeed. But if that's more elegant to do with a (public) `.browser_exists()` method, then that's fine of course. And indeed `browser_path()` can cache the result once it figured it out the first time.

Reply:
my question was about .invoke(): currently it throws on missing browser.  If I get rid of the 'raise' in the browser path function, then it won't throw anymore.  I'd feel weird to create my own subprocess exception to catch, which means I'd be left with raising a different exception, and *that* might be the API break...

Reply:
invoke() would still fail in *some* way - most probably you'll get an `AttributeError` somewhere - it can't possibly succeed without a browser. The exception being different with a missing browser isn't important, and it's a corner case anyway. So I think that's okay.